Update scripts.rst (#3403)
qgis/QGIS-Documentation: source/docs/user_manual/processing/scripts.rst

.. only:: html

   .. contents::
      :local:

There are two options for writing Processing algorithms using Python.

* Extending :class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`
* Using the @alg decorator

Within QGIS, you can use :guilabel:`Create new script` in the
:guilabel:`Scripts` menu at the top of the :guilabel:`Processing Toolbox` to
open the :guilabel:`Processing Script Editor` where you can write your code.
To simplify the task, you can start with a script template by using
:guilabel:`Create new script from template` from the same menu.
This opens a template that extends
:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`.

If you save the script in the script to the :file:`scripts` folder
(the default location) with a :file:`.py` extension, the algorithm will
become available in the :guilabel:`Processing Toolbox`.

Extending QgsProcessingAlgorithm
--------------------------------

The following code takes a vector layer, counts the number of features, does a
buffer operation and creates a raster layer from the result of the buffer
operation.
The buffer layer, raster layer and number of features are returned.

> .. test-code:: 

Author: ```suggestion
.. testcode:: 
```

>     from PyQt5.QtCore import QCoreApplication

Reviewer: ```suggestion
    from qgis.PyQt.QtCore import QCoreApplication
```

Author: Thanks for spotting that!

    
        def createInstance(self):
            return ExampleProcessingAlgorithm()
    
        def name(self):
            """
            Returns the algorithm name, used for identifying the algorithm.
            """
            return 'bufferrasterextend'
    
        def displayName(self):
            """
            Returns the translated algorithm name.
            """
            return self.tr('Buffer and export to raster (extend)')
    
        def group(self):
            """
            Returns the name of the group this algorithm belongs to.
            """
            return self.tr('Example scripts')
    
        def groupId(self):
            """
            Returns the unique ID of the group this algorithm belongs to.
            """
            return 'examplescripts'
    
        def shortHelpString(self):
            """
            Returns a localised short helper string for the algorithm.
            """
            return self.tr('Example algorithm short description')
    
        def initAlgorithm(self, config=None):
            """
            Here we define the inputs and outputs of the algorithm.
            """
            self.addParameter(

>                 QgsProcessingParameterVectorLayer(

Author: OK. Will fix.

>                 QgsProcessingParameterVectorLayer(

Author: @nyalldawson, when I change to feature source, the buffer algorithm is not happy (3.4 and 3.6).  I guess I need to prepare the feature source (convert it to something else) for the buffer algorithm.  What is the best way to do that in a processing algorithm?

                QgsProcessingOutputNumber(
                    'NUMBEROFFEATURES',
                    self.tr('NUMBEROFFEATURES')
                )
            )
    
        def processAlgorithm(self, parameters, context, feedback):
            """
            Here is where the processing itself takes place.
            """
            inputlayer = self.parameterAsVectorLayer(parameters, 'INPUTVECTOR',
                                                     context)
            numfeatures = inputlayer.featureCount()
            bufferlayer = self.parameterAsOutputLayer(parameters, 'BUFFER',
                                                      context)
            outputraster = self.parameterAsOutputLayer(parameters, 'OUTPUTRASTER',
                                                       context)
            bufferdist = self.parameterAsDouble(parameters, 'BUFFERDIST', context)
            rastercellsize = self.parameterAsDouble(parameters, 'CELLSIZE',
                                                    context)
            if feedback.isCanceled():
                return {'OUTPUTRASTER': None, 'BUFFER': None,
                        'NUMBEROFFEATURES': numfeatures}

>             inpbuffer = processing.run('native:buffer',

Author: Yes.

                                   {'INPUT': inputlayer, 'OUTPUT': bufferlayer,
                                    'DISTANCE': bufferdist, 'SEGMENTS': 10, 
                                    'DISSOLVE': True, 'END_CAP_STYLE': 0,
                                    'JOIN_STYLE': 0, 'MITER_LIMIT': 10
                                    },
                                   is_child_algorithm=True, context=context,
                                   feedback=feedback)
            # both bufferlayer and inpbuffer['OUTPUT'] refer to the result
            if feedback.isCanceled():
                return {'OUTPUTRASTER': None, 'BUFFER': bufferlayer,
                        'NUMBEROFFEATURES': numfeatures}

>             rasterized = processing.run('qgis:rasterize',

Author: OK.

                                   {'LAYER': bufferlayer, 'EXTENT': bufferlayer,
                                    'MAP_UNITS_PER_PIXEL': rastercellsize,
                                    'OUTPUT': outputraster
                                   },
                                   is_child_algorithm=True, context=context,
                                   feedback=feedback)

>             # both outputraster and rasterized['OUTPUT'] refer to the result

Author: Fixed.

            return {'OUTPUTRASTER': outputraster, 'BUFFER': bufferlayer,
                    'NUMBEROFFEATURES': numfeatures}
 
The @alg decorator
------------------

By using the @alg decorator, you can create your own algorithms by writing the
corresponding Python code and adding a few extra lines to supply additional
information needed to define the semantics of the algorithm.

> 

Author: I have added some more text.

The following code takes a vector layer, counts the number of features, does a
buffer operation and creates a raster layer from the result of the buffer
operation using the @alg decorator.
The buffer layer, raster layer and number of features are returned.

> .. test-code:: 

Author: ```suggestion
.. testcode:: 
```

    import processing
    from qgis.processing import alg
    from qgis.core import QgsProject
    
    @alg(name='bufferrasteralg', label='Buffer and export to raster (alg)', group='examplescripts',
         group_label='Example scripts')

>     @alg.input(type=alg.VECTOR_LAYER, name='INPUTVECTOR', label='Input vector layer')

Author: I have tried to include this.

    @alg.input(type=alg.RASTER_LAYER_DEST, name='OUTPUTRASTER', label='OUTPUTRASTER')
    @alg.input(type=alg.VECTOR_LAYER_DEST, name='BUFFER', label='BUFFER')
    @alg.input(type=alg.DISTANCE, name='BUFFERDIST', label='BUFFER DISTANCE', default=1.0)
    @alg.input(type=alg.DISTANCE, name='CELLSIZE', label='RASTER CELL SIZE', default=10.0)
    @alg.output(type=alg.NUMBER, name='NUMBEROFFEATURES', label='NUMBEROFFEATURES')
    
    def bufferrasteralg(instance, parameters, context, feedback, inputs):
        """
        Description of the algorithm.
        (If there is no comment here, you will get an error)
        """

>         inputlayer = instance.parameterAsVectorLayer(parameters, 'INPUTVECTOR', context)

Author: Fixed.

                                   {'LAYER': bufferlayer, 'EXTENT': bufferlayer,
                                    'MAP_UNITS_PER_PIXEL': rastercellsize,
                                    'OUTPUT': outputraster
                                   },
                                   is_child_algorithm=True, context=context,
                                   feedback=feedback)
        # both outputraster and rasterized['OUTPUT'] refer to the result
        return {'OUTPUTRASTER': outputraster, 'BUFFER': bufferlayer,
                'NUMBEROFFEATURES': numfeatures}

As you can see, it involves two algorithms.
The last one creates a raster from the buffer layer that was
generated by the first one.

The part of the code where this processing takes place is not
difficult to understand if you have read the previous chapter.
The first lines, however, need some additional explanation.
They provide the information that is needed to turn your code into
an algorithm that can be run from any of the GUI components, like
the toolbox or the graphical modeler.

These lines are all calls to the ``@alg`` functions that help
simplify the coding of the algorithm.

> The @alg function is used to define the name and location of the

Author: Yes.

algorithm in the Toolbox.

> The @alg.input function is used to define the inputs of the algorithm.

Author: OK.

The @alg.output function is used to define the outputs of the algorithm.

Here is the list of input and output types types that are supported in

> processing and their corresponding alg decorator constants

Author: Thanks.

   * - :class:`QgsProcessingParameterBoolean <qgis.core.QgsProcessingParameterBoolean>`
     - ``alg.BOOL``
     - A boolean value.
   * - :class:`QgsProcessingParameterCrs <qgis.core.QgsProcessingParameterCrs>`
     - ``alg.CRS``
     - A Coordinate Reference System
   * - :class:`QgsProcessingParameterDistance <qgis.core.QgsProcessingParameterDistance>`
     - ``alg.DISTANCE``
     - A double numeric parameter for distance values.
   * - :class:`QgsProcessingParameterEnum <qgis.core.QgsProcessingParameterEnum>`
     - ``alg.ENUM``
     - An enumeration, allowing for selection from a set of predefined values.
   * - :class:`QgsProcessingParameterExpression <qgis.core.QgsProcessingParameterExpression>`
     - ``alg.EXPRESSION``
     - An expression.
   * - :class:`QgsProcessingParameterExtent <qgis.core.QgsProcessingParameterExtent>`
     - ``alg.EXTENT``
     - A spatial extent defined by xmin, xmax, ymin, ymax.
   * - :class:`QgsProcessingParameterField <qgis.core.QgsProcessingParameterField>`
     - ``alg.FIELD``
     - A field in the attribute table of a vector layer.
   * - :class:`QgsProcessingParameterFile <qgis.core.QgsProcessingParameterFile>`
     - ``alg.FILE``

>      - A filename.

Author: Fixed.

   * - :class:`QgsProcessingParameterFileDestination <qgis.core.QgsProcessingParameterFileDestination>`
     - ``alg.FILE_DEST``

>      - A filename.

Author: Fixed.

     - A point.
   * - :class:`QgsProcessingParameterRange <qgis.core.QgsProcessingParameterRange>`
     - ``alg.RANGE``
     - A number range.
   * - :class:`QgsProcessingParameterRasterLayer <qgis.core.QgsProcessingParameterRasterLayer>`
     - ``alg.RASTER_LAYER``
     - A raster layer.
   * - :class:`QgsProcessingParameterRasterDestination <qgis.core.QgsProcessingParameterRasterDestination>`
     - ``alg.RASTER_LAYER_DEST``
     - A raster layer.
   * - :class:`QgsProcessingParameterFeatureSink <qgis.core.QgsProcessingParameterFeatureSink>`
     - ``alg.SINK``
     - A feature sink.
   * - :class:`QgsProcessingParameterFeatureSource <qgis.core.QgsProcessingParameterFeatureSource>`
     - ``alg.SOURCE``
     - A feature source.
   * - :class:`QgsProcessingParameterString <qgis.core.QgsProcessingParameterString>`
     - ``alg.STRING``
     - A text string.
   * - :class:`QgsProcessingParameterVectorLayer <qgis.core.QgsProcessingParameterVectorLayer>`
     - ``alg.VECTOR_LAYER``
     - A vector layer.
   * - :class:`QgsProcessingParameterVectorDestination <qgis.core.QgsProcessingParameterVectorDestination>`
     - ``alg.VECTOR_LAYER_DEST``
     - A vector layer.

> Layer values are strings containing the filepath of the corresponding

Reviewer 2: This isn't correct - I'd totally drop this paragraph. Instead, the `parameterAs...` methods should ALWAYS be used.

Author: Removed.

object.
A multiple layer input will come as a string value, containing the filepaths
to all selected objects, separated by semicolons (``;``).

Handing data produced by the algorithm
--------------------------------------

When you declare an output representing a layer (raster or vector),
the algorithm will try to add it to QGIS once it is finished.

* Raster layer output: QgsProcessingParameterRasterDestination /
  alg.RASTER_LAYER_DEST.
* Vector layer output: QgsProcessingParameterVectorDestination /
  alg.VECTOR_LAYER_DEST.

> So even if the ``run()`` method does not load the layers it produces,

Reviewer 2: This is confusing -- I'd clarify by saying

"does not add the layers it creates into the user's current project,..."

Author: Fixed.

the two output layers (buffer and raster buffer) will be loaded,
since they are saved to the destinations entered by the user (or to
temporary destinations if the user does not specify a destination).

> Do not use the ``load()`` method in your script algorithms.

Reviewer 2: Which "load" method does this refer to?

Author: processing.load.  I have updated the text, but please have a second look!

Reviewer 2: Argh - processing.load is old code -- that should have been removed with 3.0. Definetly ALL mentions of processing.load should be removed.

IMO, the ONLY processing.* methods which should be used are:
- algorithmHelp
- run
- runAndLoadResults
- createAlgorithmDialog
- execAlgorithmDialog
- createContext
- isWindows
- isMac

Nothing else should be used, and it's all going to be hidden away in 4.0.

If a layer is created as output of an algorithm, it should be
declared as such.
Otherwise, you will not be able to properly use the algorithm in the
modeler, since what is declared will not match what the algorithm
really creates.

Hidden outputs (numbers and strings) do not have a value. Instead, it is you who
has to assign a value to them. To do so, just set the value of a variable with
the name you used to declare that output. For instance, if you have used this
declaration,

> You can return strings, numbers and more by specifying them in the result

Reviewer 2: Better state that these need to be explicitly defined as outputs from your algorithm upfront.

I'd also say something like: "We encourage algorithms to output as many useful values as possible, since these can be valuable for use in later algorithms when your algorithm is used as part of a model"

Author: I have tried to update according to your suggestion.

dictionary (as demonstrated for "NUMBEROFFEATURES").

Communicating with the user
---------------------------

If your algorithm takes a long time to process, it is a good idea to inform the
user about the progress.  You can use ``feedback``
(:class:`QgsProcessingFeedback <qgis.core.QgsProcessingFeedback>`) for this.

The progress text and progressbar can be updated using two methods:
:meth:`setProgressText(text) <qgis.core.QgsProcessingFeedback.setProgressText>`

> and :meth:`setProgress(percent) <qgis.core.QgsProcessingFeedback.setProgress>`.

Author: ```suggestion
and :meth:`setProgress(percent) <qgis.core.QgsFeedback.setProgress>`.
```

                    self.tr('BUFFER'),
                )
            )
            # 'OUTPUT' is the recommended name for the main output parameter
            self.addParameter(
                QgsProcessingParameterRasterDestination(
                    'OUTPUT', # 'OUTPUT' is recommended for the main input
                    self.tr('OUTPUT')
                )
            )
            self.addParameter(
                QgsProcessingParameterDistance(
                    'BUFFERDIST',
                    self.tr('BUFFERDIST'),
                    defaultValue = 1
                )
            )
            self.addParameter(
                QgsProcessingParameterDistance(
                    'CELLSIZE',
                    self.tr('CELLSIZE'),
                    defaultValue = 10
                )
            )
            self.addOutput(
                QgsProcessingOutputNumber(
                    'NUMBEROFFEATURES',
                    self.tr('NUMBEROFFEATURES')
                )
            )
    
        def processAlgorithm(self, parameters, context, feedback):
            """
            Here is where the processing itself takes place.
            """

>             inputlayer = self.parameterAsSource(parameters, 'INPUT',

Reviewer 2: ```suggestion
            input_features = self.parameterAsSource(parameters, 'INPUT',
```

I'd steer away from labeling this a layer (as it's not!) -- the api is different and only part of the QgsVectorLayer api is available.

                                                     context)
            numfeatures = inputlayer.featureCount()

>             bufferlayer = self.parameterAsOutputLayer(parameters, 'BUFFER',

Reviewer 2: ```suggestion
            buffer_layer_path = self.parameterAsOutputLayer(parameters, 'BUFFER',
```

                                                      context)

>             outputraster = self.parameterAsOutputLayer(parameters, 'OUTPUT',

Reviewer 2: ```suggestion
            output_raster_path = self.parameterAsOutputLayer(parameters, 'OUTPUT',
```

     - ``alg.SOURCE``
     - A feature source.
   * - :class:`QgsProcessingParameterString <qgis.core.QgsProcessingParameterString>`
     - ``alg.STRING``
     - A text string.
   * - :class:`QgsProcessingParameterVectorLayer <qgis.core.QgsProcessingParameterVectorLayer>`
     - ``alg.VECTOR_LAYER``
     - A vector layer.
   * - :class:`QgsProcessingParameterVectorDestination <qgis.core.QgsProcessingParameterVectorDestination>`
     - ``alg.VECTOR_LAYER_DEST``
     - A vector layer.

Handing data produced by the algorithm
--------------------------------------

When you declare an output representing a layer (raster or vector),
the algorithm will try to add it to QGIS once it is finished.

* Raster layer output: QgsProcessingParameterRasterDestination /
  alg.RASTER_LAYER_DEST.
* Vector layer output: QgsProcessingParameterVectorDestination /
  alg.VECTOR_LAYER_DEST.

So even if the ``processing.run()`` method does not add the layers
it creates into the user's current project,
the two output layers (buffer and raster buffer) will be loaded,
since they are saved to the destinations entered by the user (or to
temporary destinations if the user does not specify a destination).

> Do not use the ``processing.load()`` method in your script algorithms.

Reviewer 2: I'd completely remove this paragraph - processing.load should not be used in any circumstance

Author: Removed.

:meth:`pushInfo(text) <qgis.core.QgsProcessingFeedback.pushInfo>` and
:meth:`reportError(text) <qgis.core.QgsProcessingFeedback.reportError>`.

If your script has a problem, the correct way of handling it is to raise
a :class:`QgsProcessingException <qgis.core.QgsProcessingException>`.
You can pass a message as an argument to the constructor of the exception.
Processing will take care of handling it and communicating with the user,
depending on where the algorithm is being executed from (toolbox, modeler,
Python console, ...)

Documenting your scripts
------------------------

You can document your scripts by overloading the
:meth:`helpString() <qgis.core.QgsProcessingAlgorithm.helpString>` and
:meth:`helpUrl() <qgis.core.QgsProcessingAlgorithm.helpUrl>` methods of
:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`.

Flags
-----

You can override the :meth:`flags <qgis.core.QgsProcessingAlgorithm.flags>` method of
:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`
to tell QGIS more about your algorithm.
You can for instance tell QGIS that the script shall be hidden from
the modeler, that it can be canceled, that it is not thread safe,

> and more.

Reviewer 2: Hint: By default, Processing runs algorithms in a separate thread in order to keep QGIS responsive while the processing task runs. If your algorithm is regularly crashing, you are probably using API calls which are not safe to do in a background thread. Try returning the QgsProcessingAlgorithm.FlagNoThreading flag from your algorithm's flags() method to force Processing to run your algorithm in the main thread instead.

Author: Added.

            self.addParameter(
                QgsProcessingParameterRasterDestination(
                    'OUTPUT', # 'OUTPUT' is recommended for the main input
                    self.tr('OUTPUT')
                )
            )
            self.addParameter(
                QgsProcessingParameterDistance(
                    'BUFFERDIST',
                    self.tr('BUFFERDIST'),
                    defaultValue = 1
                )
            )
            self.addParameter(
                QgsProcessingParameterDistance(
                    'CELLSIZE',
                    self.tr('CELLSIZE'),
                    defaultValue = 10
                )
            )
            self.addOutput(
                QgsProcessingOutputNumber(
                    'NUMBEROFFEATURES',
                    self.tr('NUMBEROFFEATURES')
                )
            )
    
        def processAlgorithm(self, parameters, context, feedback):
            """
            Here is where the processing itself takes place.
            """
            input_featuresource = self.parameterAsSource(parameters, 'INPUT',
                                                     context)

>             numfeatures = inputlayer.featureCount()

Reviewer 2: ```suggestion
            numfeatures = input_featuresource.featureCount()
```

            buffer_layer_path = self.parameterAsOutputLayer(parameters, 'BUFFER',
                                                      context)
            output_raster_path = self.parameterAsOutputLayer(parameters, 'OUTPUT',
                                                       context)
            bufferdist = self.parameterAsDouble(parameters, 'BUFFERDIST', context)
            rastercellsize = self.parameterAsDouble(parameters, 'CELLSIZE',
                                                    context)
            if feedback.isCanceled():
                return {'OUTPUT': None, 'BUFFER': None,
                        'NUMBEROFFEATURES': numfeatures}
            buffer_result = processing.run('native:buffer',

>                                    {'INPUT': inputlayer, 'OUTPUT': bufferlayer,

Reviewer 2: ```suggestion
                                   {'INPUT': input_featuresource, 'OUTPUT': buffer_layer_path,
```

                                    'DISTANCE': bufferdist, 'SEGMENTS': 10, 
                                    'DISSOLVE': True, 'END_CAP_STYLE': 0,
                                    'JOIN_STYLE': 0, 'MITER_LIMIT': 10
                                    },
                                   is_child_algorithm=True, context=context,
                                   feedback=feedback)
            if feedback.isCanceled():
                return {'OUTPUT': None, 'BUFFER': buffer_result['OUTPUT'],
                        'NUMBEROFFEATURES': numfeatures}
            rasterized_result = processing.run('qgis:rasterize',

>                                    {'LAYER': bufferlayer, 'EXTENT': bufferlayer,

Reviewer 2: ```suggestion
                                   {'LAYER': bufferlayer, 'EXTENT': buffer_result['OUTPUT'],
```

Author: If you say so... :-)

                                    'MAP_UNITS_PER_PIXEL': rastercellsize,

>                                     'OUTPUT': outputraster

Reviewer 2: ```suggestion
                                    'OUTPUT': output_raster_path
```

:meth:`pushInfo(text) <qgis.core.QgsProcessingFeedback.pushInfo>` and
:meth:`reportError(text) <qgis.core.QgsProcessingFeedback.reportError>`.

If your script has a problem, the correct way of handling it is to raise
a :class:`QgsProcessingException <qgis.core.QgsProcessingException>`.
You can pass a message as an argument to the constructor of the exception.
Processing will take care of handling it and communicating with the user,
depending on where the algorithm is being executed from (toolbox, modeler,
Python console, ...)

Documenting your scripts
------------------------

You can document your scripts by overloading the
:meth:`helpString() <qgis.core.QgsProcessingAlgorithm.helpString>` and
:meth:`helpUrl() <qgis.core.QgsProcessingAlgorithm.helpUrl>` methods of
:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`.

Flags
-----

You can override the :meth:`flags <qgis.core.QgsProcessingAlgorithm.flags>` method of
:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`
to tell QGIS more about your algorithm.
You can for instance tell QGIS that the script shall be hidden from
the modeler, that it can be canceled, that it is not thread safe,
and more.

> .. hint::

Reviewer: I don't know if it's a usual directive in Sphinx but how about using `..note::` or .`. tip::` instead?

Author: hint is listed as a directive in the rst documentation.
http://docutils.sourceforge.net/docs/ref/rst/directives.html#admonitions

Reviewer: OK. But is it worth adding a new directive we never used while we can reuse existing ones? Also note that the ones we used have [their own formatting](https://github.com/qgis/QGIS-Documentation/blob/master/themes/qgis-theme/static/qgis-style.less) unlike the hint one which may fall back to the "default" admonition formatting (which I don't know what it looks like btw)

Author: I used "hint" because that was the word that @nyalldawson used.  But I will change - probably best with "tip"?

> .. hint::

Author: ```suggestion
.. tip::
```

   |updatedisclaimer|

Writing new Processing algorithms as Python scripts
===================================================

.. only:: html

   .. contents::
      :local:

There are two options for writing Processing algorithms using Python.

* Extending :class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`
* Using the @alg decorator

Within QGIS, you can use :guilabel:`Create new script` in the
:guilabel:`Scrips` menu at the top of the :guilabel:`Processing Toolbox` to
open the :guilabel:`Processing Script Editor` where you can write your code.
To simplify the task, you can start with a script template by using
:guilabel:`Create new script from template` from the same menu.
This opens a template that extends
:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`.

> When saving the script to the :file:`scripts` folder (the default location)

Author: Yes, I think so.

> When saving the script to the :file:`scripts` folder (the default location)

Reviewer: Then what about formulating it as an action to do ie
```
Save the script to the script folder _(could we have the full path ? I guess it has to do with user-profiles sub folders)_
with a :file:`.py` extension and the algorithm will be made available in the :guilabel:`Processing Toolbox`
```

> When saving the script to the :file:`scripts` folder (the default location)

Author: I have updated the sentence.

                                    'DISSOLVE': True, 'END_CAP_STYLE': 0,
                                    'JOIN_STYLE': 0, 'MITER_LIMIT': 10
                                    },
                                   is_child_algorithm=True, context=context,
                                   feedback=feedback)
        # both bufferlayer and inpbuffer['OUTPUT'] refer to the result
        if feedback.isCanceled():
            return {'OUTPUTRASTER': None, 'BUFFER': bufferlayer,
                    'NUMBEROFFEATURES': numfeatures}
        rasterized = processing.run('qgis:rasterize',
                                   {'LAYER': bufferlayer, 'EXTENT': bufferlayer,
                                    'MAP_UNITS_PER_PIXEL': rastercellsize,
                                    'OUTPUT': outputraster
                                   },
                                   is_child_algorithm=True, context=context,
                                   feedback=feedback)
        # both outputraster and rasterized['OUTPUT'] refer to the result
        return {'OUTPUTRASTER': outputraster, 'BUFFER': bufferlayer,
                'NUMBEROFFEATURES': numfeatures}

> As you can see, it involves 3 algorithms, all of them coming from SAGA. The last

Reviewer: SAGA? Is this still valid with the changes you made?

Author: OOps...

Let's have the following code, which calculates the Topographic Wetness Index
(TWI) directly from a DEM

> .. note:: The simplified way of writing Python scripts (as used in the example

Author: ```suggestion
.. note:: A simplified way of writing Python scripts (as demonstrated using the syntax of QGIS 2 in the example
```

>    below) is not supported in QGIS 3.

Reviewer: Sorry I don't understand the logic: we update the code but it's not compatible with 3. What does it need to be compatible? Maybe a dev can help us fix it right now instead of reporting this fix to later (and forgetting)

Author: I have asked if https://github.com/qgis/QGIS/pull/8586 could be included in 3.4.  If it can, updating this section would be a lot easier.  The current situation is that there is no "easy" way to creating Python scripts in 3.4, in contrast to QGIS 2.  The https://github.com/qgis/QGIS/pull/8586 way (merged in master) is not the same as the QGIS 2 way, but much better than extending QgsProcessingAlgorithm.
So this is a (temporary) way to indicate that the code is only partly updated to QGIS 3.

Author: The discussion in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/134 seems to have ended - qgis/QGIS#8586 will probably not be included in 3.4, so this updating this section will require some effort.

Reviewer: @havatv I guess we can revive this PR for master now

information that is needed to turn your code into an algorithm that can be run from any
of the GUI components, like the toolbox or the graphical modeler.

These lines are all calls to the ``@alg`` functions that help simplify the coding of
the algorithm.

The @alg function is used to define the name and location of the algorithm in the Toolbox.
The @alg.input function is used to define the inputs of the algorithm.
The @alg.output function is used to define the outputs of the algorithm.

Here is the list of input and output types types that are supported in
processing and their corresponding alg decorator constants
(:file:`algfactory.py` contains the complete list of alg constants).

+---------------------------------------------------+---------------------------+--------------------+
| Class                                             | alg                       |                    |
+===================================================+===========================+====================+

> | :class:`QgsProcessingParameterAuthConfig`         | ``alg.AUTH_CFG``          |                    |

Reviewer: What could be nice is to point to each class doc in the PyQGIS API, ie using 
```
:class:`QgsProcessingParameterAuthConfig <qgis.core.QgsProcessingParameterAuthConfig>`
```
Also it would be nice to use a simpler table format (this one is hard to maintain - you'll see with my above suggestion 😈 ), maybe a list table would better fit here.

information that is needed to turn your code into an algorithm that can be run from any
of the GUI components, like the toolbox or the graphical modeler.

These lines are all calls to the ``@alg`` functions that help simplify the coding of
the algorithm.

The @alg function is used to define the name and location of the algorithm in the Toolbox.
The @alg.input function is used to define the inputs of the algorithm.
The @alg.output function is used to define the outputs of the algorithm.

Here is the list of input and output types types that are supported in
processing and their corresponding alg decorator constants
(:file:`algfactory.py` contains the complete list of alg constants).

> .. list-table:: Input and output types

Reviewer: I'm wondering if there's a better place we could add this (actually, whether we could feed the algorithm parameters description with what we have here

Author: Not sure either, but it is nice to have it available somewhere.

     - A raster layer.
   * - :class:`QgsProcessingParameterFeatureSink <qgis.core.QgsProcessingParameterFeatureSink>`
     - ``alg.SINK``
     - A feature sink.
   * - :class:`QgsProcessingParameterFeatureSource <qgis.core.QgsProcessingParameterFeatureSource>`
     - ``alg.SOURCE``
     - A feature source.
   * - :class:`QgsProcessingParameterString <qgis.core.QgsProcessingParameterString>`
     - ``alg.STRING``
     - A text string.
   * - :class:`QgsProcessingParameterVectorLayer <qgis.core.QgsProcessingParameterVectorLayer>`
     - ``alg.VECTOR_LAYER``
     - A vector layer.
   * - :class:`QgsProcessingParameterVectorDestination <qgis.core.QgsProcessingParameterVectorDestination>`
     - ``alg.VECTOR_LAYER_DEST``
     - A vector layer.

Layer values are strings containing the filepath of the corresponding
object.
A multiple layer input will come as a string value, containing the filepaths
to all selected objects, separated by semicolons (``;``).

Handing data produced by the algorithm
--------------------------------------

When you declare an output representing a layer (raster or vector),
the algorithm will try to add it to QGIS once it is finished.

> Raster layer output: QgsProcessingParameterRasterDestination /

Reviewer: Should these be part of a bulleted list? Otherwise the display is not really legible

Author: I have created a bullet list.

alg.RASTER_LAYER_DEST.
Vector layer output: QgsProcessingParameterVectorDestination /
alg.VECTOR_LAYER_DEST.
So even if the ``run()`` method does not load the layers it produces,
the two output layers (buffer and raster buffer) will be loaded,
since they are saved to the destinations entered by the user (or to
temporary destinations if the user does not specify a destination).

Do not use the ``load()`` method in your script algorithms.
If a layer is created as output of an algorithm, it should be
declared as such.
Otherwise, you will not be able to properly use the algorithm in the

> modeler, since what is declared will not not match what the algorithm

Reviewer: ```suggestion
modeler, since what is declared will not match what the algorithm
```
I didn't check where we are in the file but is it modeler or script?

Author: The modeller depends on the declaration of the parameters.

really creates.

Hidden outputs (numbers and strings) do not have a value. Instead, it is you who
has to assign a value to them. To do so, just set the value of a variable with
the name you used to declare that output. For instance, if you have used this
declaration,

You can return strings, numbers and more by specifying them in the result
dictionary (as demonstrated for "NUMBEROFFEATURES").

Communicating with the user
---------------------------

If your algorithm takes a long time to process, it is a good idea to inform the
user about the progress.  You can use ``feedback``
(:class:`QgsProcessingFeedback <qgis.core.QgsProcessingFeedback>`) for this.

The progress text and progressbar can be updated using two methods:

> ``setProgressText(text)`` and ``setProgress(percent)``.

Reviewer: Can we use :meth: references for them
```suggestion
:meth:`setProgressText(text) <qgis.core.QgsProcessingFeedback.setProgressText>`
and :meth:`setProgress(percent) <qgis.core.QgsProcessingFeedback.setProgress>`.
```
I don't know if we need to indicate the data type in the brackets.

Author: I have not changed this. Should not hurt?

You can provide more information to the user using ``pushConsoleInfo(text)``,
``pushCommandInfo(text)``, ``pushDebugInfo(text)``, ``pushInfo(text)``,  and

> ``reportError(text)``.

Reviewer: same here, using methods at https://qgis.org/pyqgis/master/core/QgsProcessingFeedback.html

Author: Fixed.

> If your script has a problem, the correct way of propagating it is to raise

Reviewer: the correct way of **propagate**? I am confused by the "propagate" word...

Author: Either "the correct way of propagating" or "the correct way to propagate".

Reviewer: I meant: why would I want to **propagate the problem**? (unless I have a partial understanding of "propagate")

Author: If you are not able to handle the exception, you should pass it on to signal that something is wrong.

Reviewer 2: "the correct way of handling it" ?

Author: OK.

a :class:`QgsProcessingException <qgis.core.QgsProcessingException>`.
You can pass a message as an argument to the constructor of the exception.
Processing will take care of handling it and communicating with the user,
depending on where the algorithm is being executed from (toolbox, modeler,
Python console, ...)

Documenting your scripts
------------------------

> Your can document your scripts by overloading the ``helpString`` and

Reviewer: ```suggestion
You can document your scripts by overloading the :meth:`helpString() <qgis.core.QgsProcessingAlgorithm.helpString>` and
```

> ``helpUrl`` methods of

Reviewer: ```suggestion
:meth:`helpUrl() <qgis.core.QgsProcessingAlgorithm.helpUrl>` methods of
```

:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`.

Flags
-----

> You can override the ``flag`` method of

Reviewer: ```suggestion
You can override the :meth:`flags <qgis.core.QgsProcessingAlgorithm.flags>` method of
```
There does not seem to be a flag method though there's a flag class in [that page](https://qgis.org/pyqgis/master/core/QgsProcessingAlgorithm.html) (I don't really how to read that file)

Author: There is a "flags" method (not a "flag" method) - thanks for pointing it out!

:class:`QgsProcessingAlgorithm <qgis.core.QgsProcessingAlgorithm>`
to tell QGIS more about your algorithm.
You can for instance tell QGIS that the script shall be hidden from

> the modeller, that it can be cancelled, that it is not thread safe,

Reviewer: ```suggestion
the modeler, that it can be canceled, that it is not thread safe,
```
